### PR TITLE
87 create minimax alg

### LINKED ISSUE #87: Create minimax alg
This will allow the engine to "think" about what the next best move is.

### PR DESCRIPTION
**Pull request by abrahimt**
Added a basic minimax alg and a test for it.

### CODE REVIEW

File: engine/src/engine.rs [hunk 1]
```diff
@@ -35,6 +35,55 @@ pub fn evaluate_position(board: &Chessboard, is_white: bool) -> f64 {
...
+    };
+    let mut best_move = None;
+
+    for from in 0..64 {
+        if let Some(piece) = game_state.board.piece_at_position(from) {
+            if (piece.is_ascii_uppercase() && !game_state.white_turn)
+                || (piece.is_ascii_lowercase() && game_state.white_turn)
+            {
+                continue;
+            }
```
Comment (chanson02) [inline engine/src/engine.rs:63]:
```suggestion
            if (piece.is_ascii_uppercase() != game_state.white_turn) { continue; }
```

File: engine/src/engine.rs [hunk 2]
```diff
@@ -35,6 +35,55 @@ pub fn evaluate_position(board: &Chessboard, is_white: bool) -> f64 {
...
+
+            let possible_moves = position::active_squares(game_state.possible_moves(from));
+            for to in possible_moves {
+                let mut cloned_state = game_state.clone();
+                if cloned_state.move_piece_legally(from, to).is_ok() {
```
Comment (chanson02) [inline engine/src/engine.rs:68]: Can you flip this condition and `continue` so we can eliminate a layer of indentation?
```suggestion
                if !cloned_state.move_piece_legally(from, to).is_ok() { continue; }
```

File: engine/src/engine.rs [hunk 3]
```diff
@@ -100,6 +149,19 @@ mod tests {
...
+    fn test_minimax_best_move() {
+        let mut game_state = GameState::new();
+        let (_, best_move) = minimax(&mut game_state, 2, true);
+        assert!(best_move.is_some(), "Expected a valid best move");
+        let (from, to) = best_move.unwrap();
+        println!(
+            "Best move: {} -> {}",
+            position::square_to_string(from),
+            position::square_to_string(to)
+        );
```
Comment (chanson02) [inline engine/src/engine.rs:162]: This test should fail gracefully with a nice error message. Also, we don't need to gunk up stdio.

File: engine/src/engine.rs [hunk 4]
```diff
@@ -100,6 +149,19 @@ mod tests {
         );
     }
 
+    #[test]
```
Comment (chanson02) [inline engine/src/engine.rs:152]: We should maybe add a test where we give it a board that can mate in 2 or 3 moves and ensure that the algorithm finds that path.

Another good idea would be to test what would happen if we give it a board that's already been mated and an empty board.

File: engine/src/engine.rs [hunk 5]
```diff
@@ -35,6 +35,55 @@ pub fn evaluate_position(board: &Chessboard, is_white: bool) -> f64 {
     }
 }
 
+/// Minimax algorithm to determine the best move.
+#[allow(dead_code)]
+pub fn minimax(
```
Comment (chanson02) [inline engine/src/engine.rs:40]: Can you include an explanation of the return types in a docstring? I'm also not convinced this function belongs in the `engine.rs` file.

**Review by chanson02 — CHANGES_REQUESTED** [review]
